### PR TITLE
NFA parser for mbe matcher

### LINKED ISSUE #4777: Incorrect truncated macro expansion
This code results in a wrong expansion of the macro, where the resulting enum only has the first specified variant:

```rust
macro_rules! enum_with_unknown {
    (
        enum $name:ident {
            $(
              $variant:ident $(,)*
            ),*
        }
    ) => {
        enum $name {
            $(
              $variant,
            )*
        }
    }
}

enum_with_unknown! {
    enum ControlOpcode {
        ConnectionUpdateReq,
        ChannelMapReq,
    }
}
```

Resulting code according to "Expand macro recursively":

```rust
enum ControlOpcode {
  ConnectionUpdateReq,
}
```

(EDIT: Reduced)

### PR DESCRIPTION
**Pull request by edwin0cheng**
Almost straight porting from rustc one, but a little bit slow :(

```
rust-analyzer analysis-stats -q . 
```

From:
```log
Database loaded:     636.11ms, 277minstr
  crates: 36, mods: 594, decls: 11527, fns: 9017
Item Collection:     10.99s, 60ginstr
  exprs: 249618, ??ty: 2699 (1%), ?ty: 2101 (0%), !ty: 932
Inference:           28.94s, 123ginstr
Total:               39.93s, 184ginstr
```

To:
```log
Database loaded:     630.90ms, 277minstr
  crates: 36, mods: 594, decls: 11528, fns: 9018
Item Collection:     13.70s, 77ginstr
  exprs: 249482, ??ty: 2699 (1%), ?ty: 2101 (0%), !ty: 932
Inference:           30.27s, 133ginstr
Total:               43.97s, 211ginstr
```

Fixes #4777

### CODE REVIEW

File: crates/mbe/src/parser.rs [hunk 1]
```diff
@@ -5,7 +5,72 @@ use smallvec::SmallVec;
...
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum OpDelimited<'a> {
+    Op(&'a Op),
+    Open,
+    Close,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) struct OpDelmitedIter<'a> {
```
Comment (jonas-schievink) [inline crates/mbe/src/parser.rs:21]:
```suggestion
pub(crate) struct OpDelimitedIter<'a> {
```

**Comment by matklad** [comment]
Yeah, the regression is sizable, but it's not like we can not do it. I think some of it can be clawed back by just profiling and microoptimiaton.

However, to do that, we really need a benchmark here. I ... now realize we don't have a great benchmarking story at all, but it's better to start somewhere. 

@edwin0cheng, could you do a separate PR which includes some bencmarks? I think they should be just usual `#[test]` tests (with run_slow_test check) which just use StopWatch to observe timings. The bench should live in the mbe crate. The bench should show the same impact for this PR, but it should be much easier to run.

With the benchmark in place, and  with a good-first-issuer "optimize this", I'll be more than comfortable to merge this.

**Comment by edwin0cheng** [comment]
Rebased from #7566 and here is the benchmark result :+1: 

```
before:
mbe expand macro rules: 1.21s, 4966minstr

after
mbe expand macro rules: 2.24s, 8464minstr
````

2x slower :(

**Comment by matklad** [comment]
bors r+

Could you create an issue with mentoring instructions on how to run the benchmark? I wonder if we can nerd-snipe someone to look at the code with a profiler...

**Comment by edwin0cheng** [comment]
> Could you create an issue with mentoring instructions on how to run the benchmark? I wonder if we can nerd-snipe someone to look at the code with a profiler...

#7857

**Pull request merged**